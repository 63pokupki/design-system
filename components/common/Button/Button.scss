$ButtonFontSize: 12px;
$ButtonFontSizeIcon: 14px;
$ButtonFontWeight: 600;
$ButtonBorderRadius: 24px;
$ButtonPadding: 0 12px 1px 12px;
$ButtonHeight: 32px;
$ButtonMinWidth: 128px;
$ButtonColorPrimary: #36a6f2;
$ButtonColorClick: #0080f6;
$ButtonColorOutlined: transparent;
$ButtonFontFamily: $font-primary;

@keyframes ButtonIsLoading {
    from {
        transform: translateX(-100%);
    }

    to {
        transform: translateX(100%);
    }
}

%ButtonIsLoadingAnimation {
    position: relative;
    overflow: hidden;

    &::after {
        animation: ButtonIsLoading 1s infinite;
        content: "";
        height: 100%;
        left: 0;
        position: absolute;
        right: 0;
        top: 0;
        transform: translateX(-100%);
        z-index: 1;
        background: linear-gradient(90deg, transparent, hsla(0, 0%, 100%, 0.3), transparent);
    }
}

/** Состояния */
%spui-Button_primary {
    border: 1px solid $ButtonColorPrimary;
    background-color: $ButtonColorPrimary;
    color: #fff;
}

%spui-Button_outline {
    border: 1px solid $ButtonColorPrimary;
    background-color: $ButtonColorOutlined;
    color: $ButtonColorPrimary;
}

%spui-Button_active {
    border: 1px solid $ButtonColorClick;
    background-color: $ButtonColorClick;
    color: #fff;
}

%spui-Button_loading {
    @extend %ButtonIsLoadingAnimation;

    cursor: unset;
}

%spui-Button_disabled {
    background-color: transparentize($ButtonColorPrimary, 0.5);
    color: #fff;
    border: 1px solid transparent;
    cursor: unset;
}

/** Базовые стили */
.spui-Button {
    @extend %reset-button;

    font-size: $ButtonFontSize;
    border-radius: $ButtonBorderRadius;
    padding: $ButtonPadding;
    font-family: $ButtonFontFamily;
    display: flex;
    align-items: center;
    justify-content: center;
    height: $ButtonHeight;
    min-width: $ButtonMinWidth;
    font-weight: $ButtonFontWeight;
    transition: background-color 0.25s ease;

    &:hover {
        transition: background-color 0.25s ease;
    }

    &.is-loading {
        @extend %spui-Button_loading;
        @extend %spui-Button_active;
    }

    &.is-disabled {
        @extend %spui-Button_disabled;
    }

    .ds-icon {
        font-size: $ButtonFontSizeIcon;
    }
}

.spui-Button_primary {
    @extend %spui-Button_primary;

    &:active {
        @extend %spui-Button_active;
    }
}

.spui-Button_outlined {
    @extend %spui-Button_outline;

    &:hover {
        @extend %spui-Button_primary;
    }

    &:active {
        @extend %spui-Button_active;
    }
}

.spui-Button_block {
    width: 100%;
}

.spui-Button_icon-right {
    .spui-Button__icon {
        margin-left: 5px;
    }
}

.spui-Button_icon-left {
    .spui-Button__icon {
        margin-right: 5px;
    }
}
