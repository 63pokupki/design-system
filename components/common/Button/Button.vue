<script>
export default {
    name: 'Button',
    functional: true,
    props: {
        /** Внешний вид кнопки - стандартная, инвертированная */
        type: {
            type: String,
            default: 'primary',
            validator: (value) => ['outlined', 'primary'].includes(value),
        },
        /** Название иконки */
        iconName: {
            type: String,
            default: '',
        },
        /** Положение иконки */
        iconPosition: {
            type: String,
            default: 'right',
            validator: (value) => ['left', 'right'].includes(value),
        },
        /** Сделать блочным элементом - занимает всю ширину */
        block: {
            type: Boolean,
            default: false,
        },
        /** Неактивное состояние - нажатия не обрабатываются */
        isDisabled: {
            type: Boolean,
            default: false,
        },
        /** Состояние загрузки - нажатия необрабатываются */
        isLoading: {
            type: Boolean,
            default: false,
        },
        /** Атрибут aria-label для доступности */
        ariaLabel: {
            type: String,
            default: 'Кнопка',
        },
    },
    render: (h, { data, props, listeners, slots }) => {
        const {
            type,
            iconName,
            iconPosition,
            block,
            isDisabled,
            isLoading,
        } = props;

        const _type = `spui-Button_${type}`;
        const _block = block && 'spui-Button_block';
        const _iconPosition =
            iconPosition && `spui-Button_icon-${iconPosition}`;

        const onClick = () => {
            if (!isLoading && listeners.click) {
                listeners.click();
            }
        };

        const slot = slots().default || 'Кнопка';

        const iconElement = (
            <i class={['ds-icon spui-Button__icon', iconName]}></i>
        );

        return (
            <button
                aria-label={props.ariaLabel}
                onClick={onClick}
                type="button"
                disabled={isDisabled}
                class={[
                    'spui-Button',
                    _type,
                    _block,
                    _iconPosition,
                    { 'is-loading': isLoading, 'is-disabled': isDisabled },
                    data.class,
                    data.staticClass,
                ]}
            >
                {iconName && iconPosition === 'left' && iconElement}
                {slot}
                {iconName && iconPosition === 'right' && iconElement}
            </button>
        );
    },
};
</script>

<style lang="scss" scoped>
@import './Button.scss';
</style>
