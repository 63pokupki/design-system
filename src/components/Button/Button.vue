<template>
  <button
    :disabled="isDisabled"
    class="spui-Button"
    :class="[_type, _block, _iconpos, {'is-loading': isLoading, 'is-disabled': isDisabled}]"
  >
    <i v-if="icon.name && icon.pos == 'left'" class="ds-icon spui-Button__icon" :class="icon.name"></i>
    <slot>Кнопка</slot>
    <i v-if="icon.name && icon.pos == 'right'" class="ds-icon spui-Button__icon" :class="icon.name"></i>
  </button>
</template>

<script>
export default {
  name: "Button",
  props: {
    type: {
      type: String,
      default: "primary",
    },
    icon: {
      type: Object,
      default: () => ({
        name: '',
        pos: '' //left/right
      }),
    },
    block: {
      type: Boolean,
      default: false,
    },
    isDisabled: {
      type: Boolean,
      default: false,
    },
    isLoading: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    _type() {
      return `spui-Button_${this.type}`;
    },
    _block() {
      return this.block ? `spui-Button_block` : "";
    },
    _iconpos() {
      return this.icon.pos ? `spui-Button_icon-${this.icon.pos}` : '';
    }
  },
  methods: {
    onClick() {
      if (!this.isLoading) {
        this.$emit("click");
      }
    },
  },
};
</script>

<style lang="scss" scoped>
@import "./Button.scss";
</style>