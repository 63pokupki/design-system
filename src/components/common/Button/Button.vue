<script>
export default {
    name: "Button",
    functional: true,
    props: {
        type: {
            type: String,
            default: "primary",
            validator: (value) => {
                return ["outlined", "primary"].indexOf(value) != -1;
            },
        },
        icon: {
            type: Object,
            default: () => ({
                name: "",
                pos: "", //left/right
            }),
        },
        block: {
            type: Boolean,
            default: false,
        },
        isDisabled: {
            type: Boolean,
            default: false,
        },
        isLoading: {
            type: Boolean,
            default: false,
        },
    },
    render: (h, { data, props, listeners, slots }) => {
        const { type, icon, block, isDisabled, isLoading } = props;

        const _type = `spui-Button_${type}`;
        const _block = block ? `spui-Button_block` : "";
        const _iconpos = icon.pos ? `spui-Button_icon-${icon.pos}` : "";

        const onClick = () => {
            if (!isLoading && listeners.click) {
                listeners.click();
            }
        };

        const slot = slots().default || "Кнопка";

        const iconElement = (
            <i class={["ds-icon spui-Button__icon", icon.name]}></i>
        );

        return (
            <button
                onClick={onClick}
                disabled={isDisabled}
                class={[
                    "spui-Button",
                    _type,
                    _block,
                    _iconpos,
                    { "is-loading": isLoading, "is-disabled": isDisabled },
                    data.class,
                    data.staticClass,
                ]}
            >
                {icon.name && icon.pos == "left" ? iconElement : null}
                {slot}
                {icon.name && icon.pos == "right" ? iconElement : null}
            </button>
        );
    },
};
</script>

<style lang="scss" scoped>
@import "./Button.scss";
</style>
