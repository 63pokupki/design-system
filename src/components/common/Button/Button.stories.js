import "@/components/resources";

import Button from "./Button.vue";

export default {
    title: "Общие/Button",
    component: Button,
};

export const empty = () => ({
    components: { Button },
    methods: {
        onClick() {
            console.log("click");
        },
    },
    template: "<Button></Button>",
});

export const primary = () => ({
    template: "<Button type='primary'>Кнопка стандартная</Button>",
    components: { Button },
    methods: {
        onClick() {
            console.log("click");
        },
    },
});

export const outlined = () => ({
    components: { Button },
    methods: {
        onClick() {
            console.log("click");
        },
    },
    template: "<Button type='outlined'>Кнопка outline</Button>",
});

export const loading = () => ({
    components: { Button },
    methods: {
        onClick() {
            console.log("click");
        },
    },
    template: "<Button :is-loading='true'>Кнопка в состоянии ожидания</Button>",
});

export const disabled = () => ({
    components: { Button },
    methods: {
        onClick() {
            console.log("click");
        },
    },
    template: "<Button :is-disabled='true'>Кнопка в неактивном состоянии</Button>",
});

export const block = () => ({
    components: { Button },
    methods: {
        onClick() {
            console.log("click");
        },
    },
    template: "<Button :block='true'>Кнопка занимает всю доступную ширину</Button>",
});

export const withIcon = () => ({
    components: { Button },
    methods: {
        onClick() {
            console.log("click");
        },
    },
    render: (h) => <Button icon={{ name: "icon-cart", pos: "right" }}>Кнопка с иконкой</Button>,
});
