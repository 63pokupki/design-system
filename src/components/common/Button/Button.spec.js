import { mount, createLocalVue } from '@vue/test-utils';
import Button from '@/components/common/Button/Button.vue';

/** Фабрика создания компонента */
const mountComponent = (context = {}) => {
    const localVue = createLocalVue();
    return mount(Button, {
        ...context,
        localVue,
    });
};

describe('Button.vue', () => {
    it('Инициализация', () => {
        const wrapper = mountComponent();
        expect(wrapper.exists()).toBeTruthy();
    });

    it('Снимок без состояния', () => {
        const wrapper = mountComponent();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Проверка изменения типа', () => {
        const wrapper = mountComponent({
            propsData: {
                type: 'outlined',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Проверка иконки', () => {
        const wrapper = mountComponent({
            propsData: {
                iconName: 'icon-cart',
                iconPosition: 'left',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Проверка блочной модели', () => {
        const wrapper = mountComponent({
            propsData: {
                block: true,
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Проверка стандартного слота', () => {
        const wrapper = mountComponent({
            slots: {
                default: 'Другой текст кнопки',
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Проверка неактивного состояния', () => {
        const fnMock = jest.fn();

        const wrapperDisabled = mountComponent({
            propsData: {
                isDisabled: true,
            },
            listeners: {
                click: fnMock,
            },
        });

        expect(wrapperDisabled.classes('is-disabled')).toBe(true);
        wrapperDisabled.trigger('click');
        expect(fnMock).not.toHaveBeenCalled();
        expect(wrapperDisabled.html()).toMatchSnapshot();

        const fnMock2 = jest.fn();

        const wrapperActive = mountComponent({
            propsData: {
                isDisabled: false,
            },
            listeners: {
                click: fnMock2,
            },
        });

        expect(wrapperActive.classes('is-disabled')).toBe(false);
        wrapperActive.trigger('click');
        expect(fnMock2).toHaveBeenCalled();
        expect(wrapperActive.html()).toMatchSnapshot();
    });

    it('Проверка состояния загрузки', () => {
        const fnMock = jest.fn();

        const wrapperDisabled = mountComponent({
            propsData: {
                isLoading: true,
            },
            listeners: {
                click: fnMock,
            },
        });

        expect(wrapperDisabled.classes('is-loading')).toBe(true);
        wrapperDisabled.trigger('click');
        expect(fnMock).not.toHaveBeenCalled();
        expect(wrapperDisabled.html()).toMatchSnapshot();

        const fnMock2 = jest.fn();

        const wrapperActive = mountComponent({
            propsData: {
                isLoading: false,
            },
            listeners: {
                click: fnMock2,
            },
        });

        expect(wrapperActive.classes('is-loading')).toBe(false);
        wrapperActive.trigger('click');
        expect(fnMock2).toHaveBeenCalled();
        expect(wrapperActive.html()).toMatchSnapshot();
    });
});
