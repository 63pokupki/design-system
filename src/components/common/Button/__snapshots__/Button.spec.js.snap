// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button.vue Проверка блочной модели 1`] = `<button aria-label="Кнопка" class="spui-Button spui-Button_primary spui-Button_block spui-Button_icon-right">Кнопка</button>`;

exports[`Button.vue Проверка изменения типа 1`] = `<button aria-label="Кнопка" class="spui-Button spui-Button_outlined spui-Button_icon-right">Кнопка</button>`;

exports[`Button.vue Проверка иконки 1`] = `<button aria-label="Кнопка" class="spui-Button spui-Button_primary spui-Button_icon-left"><i class="ds-icon spui-Button__icon icon-cart"></i>Кнопка</button>`;

exports[`Button.vue Проверка неактивного состояния 1`] = `<button aria-label="Кнопка" disabled="disabled" class="spui-Button spui-Button_primary spui-Button_icon-right is-disabled">Кнопка</button>`;

exports[`Button.vue Проверка неактивного состояния 2`] = `<button aria-label="Кнопка" class="spui-Button spui-Button_primary spui-Button_icon-right">Кнопка</button>`;

exports[`Button.vue Проверка состояния загрузки 1`] = `<button aria-label="Кнопка" class="spui-Button spui-Button_primary spui-Button_icon-right is-loading">Кнопка</button>`;

exports[`Button.vue Проверка состояния загрузки 2`] = `<button aria-label="Кнопка" class="spui-Button spui-Button_primary spui-Button_icon-right">Кнопка</button>`;

exports[`Button.vue Проверка стандартного слота 1`] = `<button aria-label="Кнопка" class="spui-Button spui-Button_primary spui-Button_icon-right">Другой текст кнопки</button>`;

exports[`Button.vue Снимок без состояния 1`] = `<button aria-label="Кнопка" class="spui-Button spui-Button_primary spui-Button_icon-right">Кнопка</button>`;
